File: src/store/mirrorGraph.ts
// ════════════════════════════════════════════════════════════════════════════
// Vérificateur acyclique pour les miroirs de dossiers (Phase 4)
//
// Pourquoi : un dossier-miroir partage le même `childBoardId` que son original.
// Si on autorisait : DossierA contient un miroir de DossierB ET DossierB contient
// un miroir de DossierA, alors entrer dans A → on voit B → entrer dans (le miroir
// de) B → on voit A → entrer dans (le miroir de) A → on voit B → ... à l'infini.
// PixiJS et le rendu boucleraient instantanément.
//
// Ce module fournit `wouldCreateMirrorCycle()` : un BFS strict sur l'arbre des
// boards (en suivant `folder.childBoardId`) qui rejette toute création susceptible
// de fermer une boucle, à n'importe quelle profondeur.
// ════════════════════════════════════════════════════════════════════════════

import { Board, CanvasFolder } from "../types";

/**
 * Renvoie `true` si placer un miroir du dossier `originalFolderId` à l'intérieur
 * du board `targetBoardId` créerait un cycle (Inception).
 *
 * Algorithme :
 *   1. On part du `childBoardId` du dossier original.
 *   2. On parcourt en BFS tous les boards atteignables en descendant dans
 *      les dossiers ET les miroirs de dossiers (qui partagent le même childBoardId).
 *   3. Si on rencontre `targetBoardId` à un moment donné → cycle confirmé.
 *
 * Cette fonction est PURE : aucune mutation du store, aucun side-effect.
 *
 * Cas pathologique de "self-mirror" : si on essaie de mettre un miroir de A
 * directement dans A (ou dans son enfant direct, ou dans n'importe quel descendant),
 * la BFS atteint `targetBoardId` immédiatement.
 */
export function wouldCreateMirrorCycle(
  boards: Board[],
  originalFolderId: string,
  targetBoardId: string,
): boolean {
  // Helpers indexés pour O(1) lookup
  const boardById = new Map<string, Board>();
  const folderById = new Map<string, CanvasFolder>();
  for (const b of boards) {
    boardById.set(b.id, b);
    for (const f of b.folders ?? []) folderById.set(f.id, f);
  }

  const original = folderById.get(originalFolderId);
  if (!original) return false; // Dossier introuvable → pas de cycle possible

  // BFS sur les boards descendants
  const visited = new Set<string>();
  const queue: string[] = [original.childBoardId];

  while (queue.length > 0) {
    const boardId = queue.shift()!;
    if (boardId === targetBoardId) return true; // ⚠️ Cycle détecté
    if (visited.has(boardId)) continue;
    visited.add(boardId);

    const board = boardById.get(boardId);
    if (!board) continue;

    // On enfile tous les childBoards (folders + folders-miroirs partageant childBoardId)
    for (const f of board.folders ?? []) {
      queue.push(f.childBoardId);
    }
  }

  return false;
}

/**
 * Renvoie le board qui contient le dossier d'id `folderId`, ou `undefined` si
 * aucun. Utilisé en amont de `wouldCreateMirrorCycle` pour identifier le board
 * cible quand on veut placer un miroir à proximité d'un dossier existant.
 */
export function findBoardContainingFolder(
  boards: Board[],
  folderId: string,
): Board | undefined {
  return boards.find((b) => (b.folders ?? []).some((f) => f.id === folderId));
}

File: crates/glucose-core/src/mirror_graph.rs
//! Vérificateur acyclique pour les miroirs de dossiers (anti-inception) — 0 dépendance.

use crate::types::Board;
use std::collections::{HashMap, HashSet, VecDeque};

/// Renvoie `true` si placer un miroir du dossier `original_folder_id` à l'intérieur
/// du board `target_board_id` créerait un cycle d'imbrication infinie (Inception).
pub fn would_create_mirror_cycle(
    boards: &[Board],
    original_folder_id: &str,
    target_board_id: &str,
) -> bool {
    let mut board_by_id = HashMap::new();
    let mut folder_by_id = HashMap::new();

    for b in boards {
        board_by_id.insert(b.id.as_str(), b);
        for f in &b.folders {
            folder_by_id.insert(f.id.as_str(), f);
        }
    }

    let original = match folder_by_id.get(original_folder_id) {
        Some(&f) => f,
        None => return false,
    };

    let mut visited = HashSet::new();
    let mut queue = VecDeque::new();
    queue.push_back(original.child_board_id.as_str());

    while let Some(board_id) = queue.pop_front() {
        if board_id == target_board_id {
            return true; // Cycle détecté !
        }
        if visited.contains(board_id) {
            continue;
        }
        visited.insert(board_id);

        if let Some(&b) = board_by_id.get(board_id) {
            for f in &b.folders {
                queue.push_back(f.child_board_id.as_str());
            }
        }
    }

    false
}

pub fn find_board_containing_folder<'a>(boards: &'a [Board], folder_id: &str) -> Option<&'a Board> {
    boards
        .iter()
        .find(|b| b.folders.iter().any(|f| f.id == folder_id))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::CanvasFolder;

    fn folder(id: &str, child_board_id: &str) -> CanvasFolder {
        CanvasFolder {
            id: id.into(),
            name: id.into(),
            color: "#60a5fa".into(),
            x: 0.0,
            y: 0.0,
            width: 100.0,
            height: 100.0,
            child_board_id: child_board_id.into(),
            mirror_of: None,
            mirror_source: None,
        }
    }

    fn board(id: &str, folders: Vec<CanvasFolder>) -> Board {
        let mut b = Board::new(id, id);
        b.folders = folders;
        b
    }

    #[test]
    fn test_detect_direct_cycle() {
        // Dossier A dans board1 mène vers board2.
        // Mettre un miroir de A dans board2 ferait board1 -> board2 -> board2 -> ...
        let b1 = board("b1", vec![folder("fA", "b2")]);
        let b2 = board("b2", vec![]);
        let boards = [b1, b2];

        assert!(would_create_mirror_cycle(&boards, "fA", "b2"));
    }

    #[test]
    fn test_detect_indirect_cycle() {
        // b1 -> fA(b2) -> fB(b3) -> fC(b1) -> boucle fermée
        let b1 = board("b1", vec![folder("fA", "b2")]);
        let b2 = board("b2", vec![folder("fB", "b3")]);
        let b3 = board("b3", vec![]);
        let boards = [b1, b2, b3];

        // Mettre un miroir de fA dans b3 fermerait la boucle
        assert!(would_create_mirror_cycle(&boards, "fA", "b3"));
        // Mettre un miroir de fB dans b1 ne crée pas de cycle vers un nouveau board neutre
        let b_neutral = board("b_neutral", vec![]);
        let boards_neutral = [
            boards[0].clone(),
            boards[1].clone(),
            boards[2].clone(),
            b_neutral,
        ];
        assert!(!would_create_mirror_cycle(
            &boards_neutral,
            "fA",
            "b_neutral"
        ));
    }
}
